x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -361,68 +367,95 @@ private void prewarmCache() {
+                            queue.add(() -> {
+                                try {
+                                    ensureOpen();
+
+                                    logger.trace("warming cache for [{}] part [{}/{}]", fileName, part, numberOfParts);
+                                    final long startTimeInNanos = statsCurrentTimeNanosSupplier.getAsLong();
+
+                                    final CachedBlobContainerIndexInput cachedIndexInput = (CachedBlobContainerIndexInput) input.clone();
+                                    final int bytesRead = cachedIndexInput.prefetchPart(part); // TODO does not include any rate limitation
+                                    assert bytesRead == cacheFile.partBytes(part);
+
+                                    logger.trace(
+                                        () -> new ParameterizedMessage(
+                                            "part [{}/{}] of [{}] warmed in [{}] ms",
+                                            part,
+                                            numberOfParts,
+                                            fileName,
+                                            TimeValue.timeValueNanos(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos).millis()
+                                        )
+                                    );
+                                } catch (Exception e) {
+                                    logger.trace(
+                                        () -> new ParameterizedMessage(
+                                            "failed to warm cache for [{}] part [{}/{}]",
+                                            fileName,
+                                            part,
+                                            numberOfParts
+                                        ),
+                                        e
+                                    );
+                                    if (e instanceof AlreadyClosedException
+                                        || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
+                                        return; // don't rethrow in case of cache eviction or directory closing
+                                    }
+                                    throw e;
+                                } finally {
+                                    if (countDown.countDown()) {
+                                        IOUtils.closeWhileHandlingException(input);
+                                    }
                                 }
-                            }
-                        });
+                            });
+                        }
                     }
                 } catch (IOException e) {
-                    logger.trace(() -> new ParameterizedMessage("failed to warm cache for [{}]", fileName), e);
+                    logger.trace(() -> new ParameterizedMessage("unable to schedule cache prewarming for [{}]", fileName), e);
```

Comment:
do we need to drain the queue and close input here?

Reply:
Right - I checked again and the only situation I see is the directory being close so it makes sense to drain the queue. I'll address that.